torch/distributions/categorical.py

```diff
@@ -93,11 +93,10 @@ def sample(self, sample_shape=torch.Size()):
     def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
-        value_shape = torch._C._infer_size(value.size(), self.batch_shape) if self.batch_shape else value.size()
-        param_shape = value_shape + (self._num_events,)
-        value = value.expand(value_shape)
-        log_pmf = self.logits.expand(param_shape)
-        return log_pmf.gather(-1, value.unsqueeze(-1).long()).squeeze(-1)
+        value = value.long().unsqueeze(-1)
+        value, log_pmf = torch.broadcast_tensors(value, self.logits)
+        value = value[..., :1]
```

Comment:
What is this line doing?

Reply:
The value has been expanded to shape `batch_shape + (n,)` but we want it to have shape `batch_shape + (1,)`. This line slices it down to that shape. Note that this is basically free because the stride is zero on the rightmost dimension.

The reason I'm using this trick is to be able to use `broadcast_tensors()` on two tensors with different target shape: `value` with shape `batch_shape + (1,)` and `logits` with target shape `batch_shape + (n,)`. It happens to work and it's cheap.